docs/references/security/rust-canister-development-security-best-practices.md

```diff
@@ -183,43 +183,157 @@ A canister could be rendered unusable so it could never be upgraded again e.g. d
 
--   Time-of-check time-of-use: checking some condition on global state before an inter-canister call and wrongly assuming it to still hold when the call returned.
+* _Double-Spending issues._ Such issues occur when a transfer is issued twice, often because of unfavorable message scheduling. For example, suppose we check if a caller is eligible for a refund and if so, transfer some refund amount to them. When the refund ledger call returns successfully, we set a flag in the canister storage indicating that the caller has been refunded. This is vulnerable to double-spending because the refund method can be called twice by the caller in parallel, in which case it is possible that the messages before issuing the transfer (including the eligibility check) are scheduled before both callbacks. A detailed explanation of this issue can be found in the [community conversation on Security Best Practices](https://www.youtube.com/watch?v=PneRzDmf_Xw&list=PLuhDt1vhGcrez-f3I0_hvbwGZHZzkZ7Ng&index=2&t=4s).   
 
 #### Recommendation
 
--   Be aware that state may change during an inter-canister call. Carefully review your code so that this kind of bugs do not occur.
+See also: "Inter-canister calls" section in [How to audit an Internet Computer canister](https://www.joachim-breitner.de/blog/788-How_to_audit_an_Internet_Computer_canister)
 
--   See also: "Inter-canister calls" section in [How to audit an Internet Computer canister](https://www.joachim-breitner.de/blog/788-How_to_audit_an_Internet_Computer_canister)
+To address issues around message ordering that can lead to bugs, one usually employs locking mechanisms to ensure that e.g. a caller (or anyone) can only execute an entire call (which involves several messages) once at a time. A simple example is also given in the [community conversation](https://www.youtube.com/watch?v=PneRzDmf_Xw&list=PLuhDt1vhGcrez-f3I0_hvbwGZHZzkZ7Ng&index=2&t=4s) mentioned above. 
+
+The locks would usually be released in the callback. That bears the risk that the lock may never be released in case the callback traps, as we discussed in [Avoid Traps after await](#avoid-traps-after-await). In Rust, there is a nice pattern to avoid this issue by using Rust's `Drop` implementation. The example code below shows how one can implement a lock per caller (`CallerGuard`) with a `Drop` implementation. From Rust CDK version 0.5.1, any local variables still go out of scope if the callback traps, so the lock on the caller is released even in that case. This pattern can be extended e.g. to work for the following use cases:
+
```

Comment:
I would move the "extensions" to after the code block to not confuse readers that these extensions might already be in the example.

Reply:
The drop is always called, even if an inter canister call never responds or responds with an error?

Reply:
> I would move the "extensions" to after the code block to not confuse readers that these extensions might already be in the example.

Agreed, that's better. done. 

> The drop is always called, even if an inter canister call never responds or responds with an error?

Yes, 'never responds' can actually not happen due to the messaging model of the IC. If a call responds with an error, normal execution of the callback proceeds and the values are dropped when they go out of scope. The most special case are the traps, which actually use that CDK feature to manually trigger the on_cleanup to ensure drop is also called.